File: chapter04/python/UnionFind.py
"""
并查集算法, 主要解决图论中的动态连通性问题
"""


class UnionFind:
    def __init__(self, n):
        self.count = n
        self.parent = list(range(n))
        self.size = [1] * n

    # 返回x的根节点
    def find(self, x):
        while self.parent[x] != x:
            self.parent[x] = self.parent[self.parent[x]]
            x = self.parent[x]
        return x

    # 将p, q连接
    def union(self, p, q):
        root_p = self.find(p)
        root_q = self.find(q)
        if root_p == root_q:
            return

        # 小树接到大树下面，较平衡
        if self.size[root_p] > self.size[root_q]:
            self.parent[root_q] = root_p
            self.size[root_p] += self.size[root_q]
        else:
            self.parent[root_p] = root_q
            self.size[root_q] += self.size[root_p]
        self.count -= 1

    # 判断p, q是否连通
    def connected(self, p, q):
        root_p = self.find(p)
        root_q = self.find(q)
        return root_p == root_q

    # 返回图中有多少个连通分量
    def count(self):
        return self.count


if __name__ == "__main__":
    uf = UnionFind(10)
    uf.union(1, 0)
    uf.union(6, 0)
    uf.union(3, 0)

    uf.union(5, 2)

    uf.union(5, 1)

File: chapter04/cpp/UnionFind.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include "limits.h"
#include <math.h>
#include <stdlib.h>

using namespace std;

class UnionFind
{
    private:
        int count;
        vector<int> parent;
        vector<int> size;
    public:
        UnionFind(int n)
        {
            parent = vector<int>(n, 1);
            for(int i=0; i<n; i++)
                parent[i] = i;
            size = vector<int>(n, 1);
            count = n;
        };

        int find(int x)
        {
            while(parent[x] != x)
                x = parent[x];
            return x;
        };

        void union_pq(int p, int q)
        {   
            int root_p = find(p);
            int root_q = find(q);
            if(root_q != root_p)
            {   
                if(size[root_p] > size[root_q])
                {
                    parent[root_q] = parent[root_p];
                    size[root_p] += size[root_q];
                }
                else
                {
                    parent[root_p] = parent[root_q];
                    size[root_q] += size[root_p];
                }
                count--;
            }
        };

        bool connected(int p, int q)
        {
            int root_p = find(p);
            int root_q = find(q);
            return root_p == root_q;  
        };

        int count_num(void)
        {
            return count;
        };
};

int main()
{
    UnionFind uf(10);
    uf.union_pq(1, 0);
    uf.union_pq(6, 0);
    uf.union_pq(3, 0);
    uf.union_pq(5, 2);
    uf.union_pq(5, 1);

    cout << uf.connected(5, 2) << endl;
    return 0;
}